Justin's Comments on Key Concepts Section of the Docs (#2255)
SFDO-Tooling/CumulusCI: docs/concepts.rst

 
**Managed packages** are typically used by Salesforce partners to distribute and sell applications to customers. They are proprietary code that can only be upgraded and deployed by the developer that built them. To ensure seamless upgrades, certain destructive changes, like removing objects or fields, cannot be performed.

Packages are also built, stored and deployed via **projects**.

Projects
--------

When you work with CumulusCI, you do so inside a project. A project is an individual git repository (think GitHub) that contains Salesforce metadata as well as the CumulusCI automation (e.g., tasks and flows) that builds and releases the project. Most importantly, a project can build one - and only one(!) - package in its repository. This one-to-one relationship between project and package is what allows version control to track any changes your team made to the repository with minimal collision.

> It's important to note that a project may not even contain a package; for example, it can be built to deliver functionality to records, or test data for QA. A project may constitute the entirety of a product made for release, or may be one of multiple projects banded together to make a product (think NPSP). Basically, whatever feature in your product that needs specialized Salesforce metadata and CumulusCI automation in order to be built and deployed, *especially* packages (we cannot stress this enough!), requires its own unique project. 

Reviewer: @davisagli I'll keep that in mind for a second run-through. There was the idea of introducing Product Delivery Model with NPSP as an example in the Introduction, but this aside can be deleted.

Project Structure
-----------------

Keychain
^^^^^^^^

Each project comes with a Salesforce CLI keychain, a unique identifier of characters to separate a project and its files from any other. This deters collisions with another org. (Example: Most projects contain a "dev" org. While the name is ubiquitous across Salesforce orgs, the keychain attached is what makes it unique.) 

Project Directory
^^^^^^^^^^^^^^^^^

The root directory of your CumulusCI project. Because each project is linked to a single GitHub repository, you can always be confident of the scope of the actions you're taking. And thanks to the keychain, CumulusCI knows which project you are working on by the current working directory of your shell. 

> [NOTE: Save yourself a headache by making sure you're in the correct repository for your project before running project-specific commands; otherwise, your project will produce an error. (This tends to be one of the biggest troubleshooting issues that developers run into with CumulusCI. Check for this first and you might save yourself an extra trip to this guide.)]

Reviewer: Helpful hint? Annotation?

Author: Dang, I didn't know we had all of those options :)

In order to be used as a CumulusCI project, a directory must both be a Git repository and contain a ``cumulusci.yml`` configuration file. We'll show you how to get set up with a new or existing CumulusCI project in the [TODO link Get Started section].

cumulusci.yml
^^^^^^^^^^^^^

The ``cumulusci.yml`` file defines a project's automation in CumulusCI. It contains all the customizations and configurations that pertain to your project's lifecycle. It can encompass everything from customizing the shapes of scratch orgs to configuring tasks and flows.

Learn more about customizing CumulusCI automation in the [TODO: link Customization section].

force-app (or src)
^^^^^^^^^^^^^^^^^^

> The main body of the project's code and metadata lives in the ``force-app`` directory for Salesforce DX-format projects and ``src`` for Metadata API-format projects. For managed package projects, only this metadata is part of the package. Projects have the vast majority of their components stored here.

Reviewer: "For managed package projects, only this metadata is part of the package."

A bit vague. Should we clarify further?

Reviewer 2: What do you think needs clarification? I think the point of the sentence is to contrast it with other sets of metadata in the unpackaged folder.

Reviewer: @davisagli I think I need a better understanding then of where metadata is stored. To start: A managed package also has an unpackaged folder?

Reviewer 2: @justinvholt It can. There's the primary package of metadata in force-app which is going to be released as a managed package, and then there can be other sets of metadata in various folders under `/unpackaged`. These can serve a variety of different purposes such as:
- occasionally it's necessary to deploy some metadata in order to enable a feature that is a prerequisite for installing a managed package
- there are some types of metadata that aren't supported in managed packages, but can still be deployed as unmanaged metadata
- there can be metadata which only needs to be deployed for particular configurations, such as some additional fields that are used for QA purposes but not wanted in other types of orgs

Reviewer: So is it that the unmanaged metadata that's part of a package is only used as a means to deploy/install the managed package but not "officially" part of (supported by?) the managed package once it's installed? (Does this question make sense?)

Is there a better way to convey how integral force-app folder/this metadata is to a managed package?

Reviewer 2: Unmanaged metadata is, by definition, not part of the managed package, but it can still be an important part of delivering a full product to customers. The end goal is not to install a package; the end goal is to deliver a fully configured product. Sometimes doing that requires deploying unmanaged metadata before or after the managed package is installed.

> Is there a better way to convey how integral force-app folder/this metadata is to a managed package?

The default package directory (usually called force-app) defines what will be included when you release a managed package from your cci project (or when you release an unlocked package, if you're doing that, or if you are not releasing the package at all but running the `deploy` task to get the metadata into an org in unmanaged form). It's the right place to put metadata except for when there's a reason to put it somewhere else (like one of the reasons I listed above).

orgs Directory
^^^^^^^^^^^^^^

> The ``.json`` files found in the ``orgs`` directory define the Salesforce DX org configurations that are available to the project. We cover scratch org management in depth in [TODO: link Scratch Org Management].

Reviewer: "We cover scratch org management in depth"

Haven't mentioned scratch orgs at this point, and there's not enough detail here (yet) that discusses the relationship between orgs directory and scratch orgs. Should we elaborate?

Reviewer 2: I wonder if we should move the info on project structure below the Project Orgs and Services and Tasks and Flows sections, since the project structure includes references to concepts introduced in those sections.

Reviewer: Let's definitely set a time to talk about this next week. There's enough of a mix of language in both sections that we might need to sort out what takes priority in getting mentioned first.

datasets
^^^^^^^^

Each project may have one or more **datasets**: on-disk representations of record data that can be inserted into Salesforce orgs, and which can also be modified and re-captured during the evolution of the project. Datasets are stored in the ``datasets`` directory. Learn more about datasets in [TODO: link Automating Data Operations].

Robot
^^^^^

> Robot Framework provides browser automation for end-to-end testing. Each project contains a ``robot`` directory, which stores the resources and test suites for the project's Robot Framework test suites. New CumulusCI projects start with a simple Robot test case that creates a Contact record.

Reviewer: "browser automation for end-to-end testing"

Does this need further clarification? For instance, what is browser automation? What is its relationship to CumulusCI?

Reviewer 2: I don't think we need much detail here. The important thing is to point out that CumulusCI includes a thing called Robot Framework which can be used to create automated browser tests, and link to the later section of the docs which covers it in detail.

> Robot Framework provides browser automation for end-to-end testing. Each project contains a ``robot`` directory, which stores the resources and test suites for the project's Robot Framework test suites. New CumulusCI projects start with a simple Robot test case that creates a Contact record.

Reviewer: "stores the resources and test suites for the project's Robot Framework test suites"

Redundant to use "test suites" twice? Can we drop the second one and just end with "Robot Framework"?

Reviewer 2: I would edit it down to simply "which stores the project's Robot Framework test suites"

> While Robot Framework is used primarily for browser automation testing, it can also be harnessed to help configure orgs where other strategies and APIs are insufficient. [TODO LINK TO EXPLAIN THIS FURTHER?]

Reviewer: TODO Link to explain this further?

Reviewer 2: We're going to fill in the links once the PRs for individual sections are merged so that there's a place to link to.

Reviewer: I figured, but I added this particular TODO, so just clarifying.

unpackaged metadata
^^^^^^^^^^^^^^^^^^^

A product doesn't just encompass the contents of a managed package or a single deployment. It also includes **unpackaged metadata**: extra bundles of Salesforce metadata that further tailor an org or complete the product.

In a CumulusCI project, all unpackaged metadata is stored in subdirectories within the ``unpackaged`` directory. Unpackaged metadata plays multiple roles, including preparing an org for installing packages, adding more customization after the package or application is deployed, and customizing specific orgs that are used in the product's development process.

Learn more about managing unpackaged metadata in [TODO: link Managing unpackaged configuration].

Project Orgs & Services
-----------------------

> Orgs and services are external, authenticated resources that each project uses. CumulusCI makes it easy to connect orgs, as well as services like GitHub or MetaDeploy, to a single project, or to use them across many projects.

Reviewer: "external, authenticated resources"

Is an org an external resource? Usually? Always? Sometimes?

Reviewer 2: Yes, it's always on a Salesforce server, external to the machine where you are running CumulusCI.

Reviewer: @davisagli Ah. See, I was interpreting external as meaning outside a Salesforce project, not the machine itself.

> Each project has its own set of orgs, including active scratch orgs, persistent orgs like a production or packaging org, and predefined scratch org configurations. CumulusCI securely stores org authentication information in its keychain, making it easy to access connected orgs at any time. The ``cci org list`` command shows all of the orgs connected to a project. Orgs can also be shared across multiple projects.

Reviewer: "predefined scratch org configurations"

Does reader automatically understand what this means? Further clarification?

Reviewer 2: Probably not, but we explain it more in the section on scratch orgs.

> Each project has its own set of orgs, including active scratch orgs, persistent orgs like a production or packaging org, and predefined scratch org configurations. CumulusCI securely stores org authentication information in its keychain, making it easy to access connected orgs at any time. The ``cci org list`` command shows all of the orgs connected to a project. Orgs can also be shared across multiple projects.

Reviewer: "Orgs can also be shared across multiple projects."

Why is this useful? How does CumulusCI assist in making this happen?

Reviewer 2: Maybe I'm working on several different projects and I want to deploy all of them into the same persistent org in the end. I can connect that persistent org as a global org, instead of the default of being specific to one project. Then I can deploy things to it from any cci project.

It's more of an advanced use case and I don't think we need to do more than mention it.

Configuring orgs in CumulusCI is powerful, but comes with some complexity. To review the details, read the sections [TODO: link Scratch org environments] and [TODO: link persistent org section].

> Services are usually, but not always, connected to CumulusCI across projects as part of the global keychain. The ``command cci service`` list shows you which services are connected in the context of the current project.

Reviewer 2: I'm not sure why you removed the Services subheading and definition of what a service is. I know it sort of repeats the first paragraph under Orgs and Services, but I think it's important to make it clear that Services are a separate concept from Orgs.

Reviewer: I believe I deleted it because it was already mentioned in the first paragraph. Will fix. (There was never a separate subheading for Org and Services but might be prudent to add one for each.)

> Services can be connected at the project level, which means that they're scoped to a single project and cannot be shared. Global services are easy to use and share and, therefore, we recommend you rely on them for the most part. However, when you encounter a scenario where, for example, you need to use a specific Dev Hub for one - and only one(!) - project, you can simply connect to that service by way of the ``cci service connect devhub --project`` command.

Reviewer: "cannot be shared"

It was originally "aren't shared". Just wanted to confirm that my wording is OK.

Reviewer 2: sure, I think they're equivalent

Tasks and Flows
---------------

CumulusCI uses a framework of **tasks** and **flows** to organize the automation that is available to each project.

Tasks are units of automation. A task could perform a deployment, load a dataset, retrieve material from an org, install a managed package, or undertake many other activities. CumulusCI ships with scores of tasks out of the box.

> You can review the tasks available in a project by running ``cci task list``; learn more about a task and how to configure its options with ``cci task info <name>``, where ``<name>`` is the name of the task; and run a task with ``cci task run <name> --org <org>``, where ``<name>`` is the name of the task and ``<org>`` is the org you'd like to run it against. For example, the ``run_tests`` task executes Apex unit tests. If you have an org called ``dev``, you can run this task against this org with the command ``cci task run run_tests --org dev``.

Reviewer: The paragraph is fine as is, but it also starts getting a little dense as one block of text. Would it be better to separate each task mentioned as its own line?

> Many operations that you'll undertake with CumulusCI, including creating new orgs, use flows. Flows are ordered sequences of tasks (and other flows!) that produce a cohesive outcome, such as an org that's configured to suit a specific workflow.

Reviewer: "specific workflow"

We might need to give an example of what this means, or turn the next paragraph into a more streamlined example(?).

Key Concepts
============

> Key concept. Developer jargon. 10-cent words that sound like a technical writer patting himself on the back. (Hi there!) We imagine your cursor subconsciously drifting its way to the hyperlink for the next chapter already. BUT in an effort to give you the tools to build and test customized features with CumulusCI, it's only right that we demystify the process by explaining the *how* and the *why* in order to help you achieve your goal of seamless automation.

Reviewer 3: I see that you're trying to ease into this topic but this approach draws too much attention to the writer's voice. There's a time and a place for that; when content is dense, it's reassuring to hear the voice of someone knowledgeable right by your side. Seeing as this content is pretty high level and general, consider a more straightforward approach. "Let's review some important concepts when building and testing features using CumulusCI." Also be vigilant for creeping jargon like "seamless." =)

Reviewer 4: In this case I am the perpetrator but my zeal for sesquipedalianism remains unbowed!

Reviewer 4: Current thinking is that this entire section moves into the Introduction, right?

Reviewer: @davidmreed Do you mean _all_ of Key Concepts moving into Introduction? As it stands, Intro is gonna be DENSE what with the explanation of CumulusCI and the Product Delivery Model.

Reviewer: @lianeyee Fiiiiiiiiiine. Writer's voice quelled... for now.

For the record, I'm gonna blame @Br4nd0R for the phrase "seamless automation". (I personally prefer my automation to be rife with seams.)

Reviewer: Entire paragraph sent out to pasture; adios, darlings. (Sir Arthur Quiller-Couch would be proud.)

Reviewer 4: I mean the Product Delivery Model section moving.

Reviewer: @davidmreed Yes, PDM will go into Intro. (Part of the reason why it's gonna be a bit of a beast to overhaul.)

> Don't worry, we'll make this as painless as possible.

Reviewer 3: If you go with the suggestion up top, this line isn't needed.

Packages
--------

> If you've come this far in your search of customization, we're guessing you're familiar with packages as well as their role in customizing a Salesforce org. However, because most everything that is built in CumulusCI is deployed via packages, it's essential to clarify once more:

Reviewer 3: ```suggestion
If you've come this far in your search for customization tools, we're guessing you're familiar with packages as well as their role in customizing a Salesforce org. However, because most everything that you build in CumulusCI is deployed via packages, it's essential to review the basics of packages.
```

> A **package** is a container for something as small as an individual component or as large as a set of related apps. After creating a package, you can distribute it to other Salesforce users and organizations, including those outside your company.

Reviewer 4: I don't think this is wrong, but the ambitious phrasing may give the wrong impression. It would be unusual for a single managed package to include a set of related apps.

Reviewer: @davidmreed What's the word for "a set of a related apps" anyway? A zynga? A benioff? 

Seriously, though, what could we use as an example for package size? "As large as a fully functional app"?

Reviewer 4: How about "... as small as an individual component or as large as a sophisticated application"?

> A **package** is a container for something as small as an individual component or as large as a set of related apps. After creating a package, you can distribute it to other Salesforce users and organizations, including those outside your company.

Reviewer 3: The typical Salesforce standard for indicating emphasis and indicating a newly introduced term is italics. We do sometimes use bold for emphasis, though not with any consistent pattern that I can detect. Flagging it for now as a formatting question for this whole doc set. I see David Reed also called out a formatting question about formatting code. We should discuss.

Reviewer: Yeah, wasn't sure if bold or italics was the way to go. I figured bold would stand out better because these are VIP key concepts.

Reviewer 3: Let's see if @jvarese has an opinion.

Reviewer 5: @justinvholt - as written, the above is definitely an italics situation (per house style). We use italics in the flow of a sentence. Another way to extra emphasize something is with a definition list, which bolds the first word/definition automatically. If this is part of a whole list of things, you could consider doing it that way too.

Reviewer: Swapped 'em out!

> **Unmanaged packages** are typically used to distribute open-source (non-proprietary) features or application templates to provide developers with the basic building blocks for an application. Once the components are installed from an unmanaged package into a specific org, it's what's known as an org implementation. These freshly installed components can be edited as seen fit by the owners of this implementation. The developer who created and uploaded the unmanaged package has no control over the installed components, and can't change or upgrade them.

Reviewer 3: ```suggestion
**Unmanaged packages** are typically used to distribute open-source (non-proprietary) features or application templates to provide developers with the basic building blocks for an application. After the components are installed from an unmanaged package in a specific org, it's what's known as an _org implementation_. These freshly installed components can be edited by the owners of the implementation. The developer who created and uploaded the unmanaged package has no control over the installed components, and can't change or upgrade them.
```

> **Managed packages** are typically used by Salesforce partners to distribute and sell applications to customers. They are proprietary code that can only be upgraded and deployed by the developer that built them. To ensure seamless upgrades, certain destructive changes, like removing objects or fields, cannot be performed.

Reviewer 3: ```suggestion
**Managed packages** are typically used by Salesforce partners to distribute and sell applications to customers. They are proprietary code that can be upgraded and deployed only by the developer that built them. To ensure seamless upgrades, we prevent certain destructive changes, such as deleting objects or fields.
```
Style point on use of "only": https://confluence.internal.salesforce.com/display/DOCTEAM/only

I'm also noticing a lot of passive voice. Edited the last sentence to make it more active. There might be a few other places to try to do the same (doesn't have to be in this round of edits).

> In CumulusCI, packages are also built, stored and deployed via **projects**.

Reviewer: @davidmreed Edit noted . . . but to clarify: packages are **not** stored via projects? Because I've been bragging about that to all my friends.

> In CumulusCI, packages are also built, stored and deployed via **projects**.

Reviewer 4: As is so often the case... it's complicated.

Projects store the source code that becomes a package. But the package itself is an artifact that is only ever stored on a Salesforce server. It's as if you had Java code but the compiled JAR file was never accessible to you.

Projects
--------

> When you work with CumulusCI, you do so inside a project. A project is an individual git repository (think GitHub) that contains Salesforce metadata as well as the CumulusCI automation (e.g., tasks and flows) that builds and releases the project. Most importantly, a project can build one - and only one(!) - package in its repository. This one-to-one relationship between project and package is what allows version control to track any changes your team made to the repository with minimal collision.

Reviewer 3: ```suggestion
When you work with CumulusCI, you do so inside a project. A project is an individual Git repository (think GitHub) that contains both Salesforce metadata and CumulusCI automation (such as tasks and flows) that builds and releases the project. Most importantly, a project can build one—and only one(!)—package in its repository. This one-to-one relationship between project and package is what allows version control to track any changes your team makes to the repository.
```
I like this explanation! Consider whether you really need the GitHub parenthetical? It seems more distracting than illuminating.

Reviewer: @lianeyee I'll remove the "Think GitHub" bit.

@davidmreed IIRC, "one-to-one relationship" came from the original text here. I discussed that in greater detail with @Br4nd0R, and that's where I came up with this description. If I'm not still not describing it accurately, what can we say that stresses the importance of the one-to-one relationship? (@davisagli, you might want to chime in here, too.)

Reviewer 2: I agree that it's inaccurate to say that the 1:1 package-repository relationship is what makes version control possible.  One package per repository is more of a convention that CumulusCI assumes in a lot of places -- it's possible to imagine storing multiple packages in the same repository and setting up custom CumulusCI flows to work with them, but you would be "working against the grain" and not able to take advantage of the default flows that CumulusCI provides. So the main point is that you'll have a better experience with CumulusCI if you accept its premise that each package should go in its own repository.

Reviewer: @davisagli Would you advise we just leave the language as is in order to "push" people to stick with the 1:1 relationship?

Reviewer 2: @justinvholt how about:

When you work with CumulusCI, you do so inside a project. A project is an individual Git repository (think GitHub) that contains both Salesforce metadata and CumulusCI automation (such as tasks and flows) that builds and releases the project. If you are building multiple packages, we strongly recommend organizing each package as a separate project in its own repository. CumulusCI's standard automation assumes that there is one package per repository, so it will work the best if you follow this convention.

> It's important to note that a project may not even contain a package; for example, it can be built to deliver functionality to records, or test data for QA. A project may constitute the entirety of a product made for release, or may be one of multiple projects banded together to make a product. Basically, whatever feature in your product that needs specialized Salesforce metadata and CumulusCI automation in order to be built and deployed, *especially* packages (we cannot stress this enough!), requires its own unique project. 

Reviewer 3: ```suggestion
A project doesn't have to contain a package. For example, it can deliver functionality to records or provide test data for QA. A project can constitute an entire product for release, or be just one of multiple projects banded together to make a product. Here's a rule of thumb for when a project _is_ required: whenever the thing that you're building and deploying requires specialized Salesforce metadata and CumulusCI automation, and *especially* packages. We can't stress this enough! 
```

Reviewer: Will definitely strip out those mays/mights, but I need to advocate here for my original text because it adds more emphasis where needed. Saying "a project doesn't have to contain a package" is succinct, but it _is_ important to note because it's the reason for this paragraph's existence.  Because projects are not just vehicles for packages. (I made this assumption several times; @Br4nd0R corrected me each time.) Also, what's being a stressed enough (at the end) isn't the fact that a project only requires SF metadata and cci automation, which had just been clarified, but that **packages** definitely require a project because they're the bread and butter of cci.

My proposed edit:

"It's important to note that a project doesn't have to contain a package. For example, it can deliver functionality to records or provide test data for QA. A project can constitute an entire product for release, or be just one of multiple projects that together constitute a product. Essentially, if whatever you're building and deploying requires specialized Salesforce metadata and CumulusCI automation, *especially* packages (we cannot stress this enough!), it requires its own unique project."

(I also prefer "cannot" over "can't" in this instance because read aloud it emphasizes the stress.)

Reviewer 3: @justinvholt Looks like we've landed in a great succinct place with David Reed's edits below and your concluding "To sum up." Yay!

> It's important to note that a project may not even contain a package; for example, it can be built to deliver functionality to records, or test data for QA. A project may constitute the entirety of a product made for release, or may be one of multiple projects banded together to make a product. Basically, whatever feature in your product that needs specialized Salesforce metadata and CumulusCI automation in order to be built and deployed, *especially* packages (we cannot stress this enough!), requires its own unique project. 

Reviewer: @lianeyee submitted her edit on this, too, so I'm tagging her here so she can have eyes on it. 

@davidmreed I really like your version because it's even more succinct. (I'll swap out the "may".) 

That said, "builds either one or zero packages" is true, but it's not a great selling point for cci. My reason for restructuring the entirety of Key Concepts is to center everything around packages, which is the primary motivator for someone to use CCI in the first place, correct? So I think it's prudent to emphasize that while a project's existence doesn't hinge on whether or not it contains a package, it _is_ the vehicle for a package.

I suggest we end this section with this:

"To sum up, although a project doesn't require a package, a package requires a project to be built and deployed."

Project Structure
-----------------

Project Directory
^^^^^^^^^^^^^^^^^

> The root directory of your CumulusCI project. Because each project is linked to a single GitHub repository, CumulusCI knows which project you are working on by the current working directory of your shell. 

Reviewer 3: I'm not sure how to read "The root directory of your CumulusCI project." Either words are missing or it should instead be a heading like "Root Directory"?

Reviewer: @davidmreed: Would "The project directory is the root of your CumulusCI project." be accurate? Would I say root directory even if we're using it twice in one sentence (yuck)? Root folder? Source?

Reviewer 4: Your phrase reads fine to me. I'd rather not use both the words "directory" and "folder" because they mean the same thing but are differently idiomatic depending on the user's background and operating system.

> .. tip:: Avoid headaches by making sure you're in the correct repository for your project before running project-specific commands; otherwise, your project will produce an error. (This tends to be one of the biggest troubleshooting issues that developers run into with CumulusCI. Check for this first and you might save yourself an extra trip to this guide.).

Reviewer 3: ```suggestion
.. tip:: Avoid headaches by making sure you're in the correct repository for your project before running project-specific commands. Otherwise, errors can happen in your project. This issue tends to be one of the biggest CumulusCI troubleshooting issues. Check your repo first and you might save yourself an extra trip to this guide.
```
Try to avoid future tense and avoid omitting a subject, like "This tends to . . ." Specify the subject, as suggested. I like the parenthetical thought but I like it better as a non-parenthetical. ;-)

Reviewer: Clean and succinct. Love it. 

That said, I prefer "project will produce an error" because that's _literally_ what will happen. The system will kick back an error, not just that an error can occur because you've saved this or that to the wrong folder.

Reviewer 3: Find a way to say that without using future tense and I'm sold. :)

> In order to be used as a CumulusCI project, a directory must both be a Git repository and contain a ``cumulusci.yml`` configuration file. We'll show you how to get set up with a new or existing CumulusCI project in the [TODO link Get Started section].

Reviewer 3: ```suggestion
In order to be used as a CumulusCI project, a directory must both be a Git repository and contain a ``cumulusci.yml`` configuration file. We cover how to get set up with a new or existing CumulusCI project in the [TODO link Get Started section].
```

> cumulusci.yml

Reviewer: @davidmreed Anything that is its own file name or command gets ``code style`` formatting. Are you talking about the phrase "standard library"? We can make that bold text.

> cumulusci.yml

Reviewer 4: No, I'm pointing at some of the headings, which are specific references to files or directories. They can be both headings and code style, I believe.

> cumulusci.yml

Reviewer: Hrm. Fair point. I defaulted on using headers without code style. But I'm not beholden to that.

@lianeyee: Thoughts on whether we should make any giving any word in a header that has its own file/folder/command code style?

> cumulusci.yml

Reviewer 3: @justinvholt If we would format a word or phrase in body text with code style, we should do the same when it appears in heading text.

> cumulusci.yml

Reviewer: Changed it, although the formatting looks a bit odd. Might need to revisit with next commit.

^^^^^^^^^^^^^

> The ``cumulusci.yml`` file defines a project's automation in CumulusCI. It contains all the customizations and configurations that pertain to your project's lifecycle. It can encompass everything from customizing the shapes of scratch orgs to configuring tasks and flows.

Reviewer 3: ```suggestion
The ``cumulusci.yml`` file defines a project's automation. It contains all the customizations and configurations that pertain to your project's lifecycle. It can encompass everything from customizing the shapes of scratch orgs to configuring tasks and flows.
```

Learn more about customizing CumulusCI automation in the [TODO: link Customization section].

force-app (or src)
^^^^^^^^^^^^^^^^^^

> The main body of the project's code and metadata lives in the default package directory, commonly shown as the ``force-app`` folder for Salesforce DX-format projects and the ``src`` folder for Metadata API-format projects. **force-app** defines what will be included when you release a managed package from your CumulusCI project (or when you release an unlocked package, or if you're not releasing the package at all but running the ``deploy`` task to get the metadata into an org in unmanaged form).

Reviewer 3: ```suggestion
The main body of the project's code and metadata lives in the default package directory, commonly shown as the ``force-app`` folder for Salesforce DX-format projects and the ``src`` folder for Metadata API-format projects. **force-app** defines what's included when you release a managed package from your CumulusCI project. Or, when you release an unlocked package, or if you're not releasing the package at all but running the ``deploy`` task to get the metadata into an org in unmanaged form.
```

Reviewer: I often use parentheticals as asides, meaning not vital but still letting the reader in the know. Again, I've written this entire page as if I'm giving a presentation aloud. So any emphasis, whether bold, italics, em-dashes or parentheses, gives a rhythm to the writing. I'm not saying this because the text is precious. But if we remove all of the parentheses, we're completely stripping away the "confidential tone" that I'm using to engage the reader.

Perhaps we can discuss with JV after the holidays and find a happy medium?

Reviewer 3: I fear that I'm giving off "I don't like parentheticals" vibes when actually, I do like them! In this case, it was more about sentence length getting long, making for more work to parse. Breaking up the sentence with a harder stop feels like it gives me permission to set down that first chunk and then consider the next chunk. The parenthetical seems to ask for more cognitive holding of that first chunk. Anyway, this is getting really esoteric and nitpicky. Just note a potential readability issue that may or may not be a real issue. (And happy to get JV's take on it.)

Reviewer: I turned the parenthetical into its own sentence. It looks a bit odd, but I still don't think it has enough weight to drop the parentheses. Additionally, it's something Parker mentioned specifically, so I'm hesitant to drop it completely because it _is_ helpful to the reader with a certain skill level.

Tagging @jvarese so he knows what we're talking about next week.

Reviewer 5: @justinvholt - I like parenthetical, though Liane's right about us needing a more definite break to the flow. Sounds like making it its own sentence is a great solution. If the semi-colon comes back after that first clause, that might help too. (And I do admire your circumspection around the parenthetical -- a difficult thing to master.)

Reviewer 5: @justinvholt - a PS: if you don't want the **forceapp** that starts the sentence to look weird, you could simply redundantly start the sentence with "The forceapp folder/directory defines . . . " I know we are always trying to write less words, not more, but this is one of those cases where need for clarity trumps that ambition.

orgs directory
^^^^^^^^^^^^^^

> The ``.json`` files found in the ``orgs`` directory define the Salesforce DX org configurations that are available to the project. We cover scratch org management in depth in [TODO: link Scratch Org Management].

Reviewer 3: ```suggestion
The ``.json`` files in the ``orgs`` directory define the Salesforce DX org configurations that are available to the project. We cover scratch org management in depth in [TODO: link Scratch Org Management].
```

datasets
^^^^^^^^

> Each project may have one or more **datasets**: on-disk representations of record data that can be inserted into Salesforce orgs, and which can also be modified and re-captured during the evolution of the project. Datasets are stored in the ``datasets`` directory. Learn more about datasets in [TODO: link Automating Data Operations].

Reviewer 3: ```suggestion
Each project can have one or more **datasets**: on-disk representations of record data that can be inserted into Salesforce orgs, and that can also be modified and re-captured during the evolution of the project. Datasets are stored in the ``datasets`` directory. Learn more about datasets in [TODO: link Automating Data Operations].
```

robot
^^^^^

> Robot Framework provides browser automation for end-to-end testing. Each project contains a ``robot`` directory, which stores the project's Robot Framework test suites. New CumulusCI projects start with a simple Robot test case that creates a Contact record.

Reviewer 3: ```suggestion
Robot Framework provides browser automation for end-to-end testing. Each project contains a ``robot`` directory, which stores the project's Robot Framework test suites. New projects start with a simple Robot test case that creates a Contact record.
```

> While Robot Framework is used primarily for browser automation testing, it can also be harnessed to help configure orgs where other strategies and APIs are insufficient. [TODO LINK TO EXPLAIN THIS FURTHER?]

Reviewer 3: ```suggestion
While Robot Framework is primarily for browser automation testing, it's also useful for configuring orgs when other strategies and APIs are insufficient. [TODO LINK TO EXPLAIN THIS FURTHER?]
```
Is it more common to say "automated browser testing" than "browser automation testing"? I didn't suggest the change because I'm not sure.

Reviewer: @davidmreed, @davisagli, @Br4nd0R: Thoughts?

Reviewer 2: I think "automated browser testing" is clearer.

unpackaged metadata
^^^^^^^^^^^^^^^^^^^

> A product doesn't just encompass the contents of a managed package or a single deployment. It also includes **unpackaged metadata**: extra bundles of Salesforce metadata that further tailor an org or complete the product.

Reviewer 3: Should this be moved up higher? We talk about metadata pretty early on in Key Concepts, and this tidbit seems helpful to know before it's first mentioned.

Reviewer: Good point. We could place it between project directory and cumulusci.yml.

@davidmreed, @davisagli, @Br4nd0R: Thoughts?

> A product doesn't just encompass the contents of a managed package or a single deployment. It also includes **unpackaged metadata**: extra bundles of Salesforce metadata that further tailor an org or complete the product.

Reviewer 3: ```suggestion
A product doesn't just include the contents of a managed package or a single deployment. It also includes **unpackaged metadata**: extra bundles of Salesforce metadata that further tailor an org or complete the product.
```
Slightly simpler.

Reviewer: Yeah, but then we're using "include" as the verb twice in a row. I think I chose encompass as shorthand for "that's not all that the project contains".

Reviewer 3: Sometimes, echoing the verb can be a desirable, helpful device that ties the second idea closely to the first. When I first read it, "include" seemed like a simpler word choice than "encompass," as well as less cognitive load--I've already processed that verb the first time. But I respect your point that "encompass" might more precisely convey what you mean. In which case, never mind my edit! Just note that repetition isn't always a bad thing.

In a CumulusCI project, all unpackaged metadata is stored in subdirectories within the ``unpackaged`` directory. Unpackaged metadata plays multiple roles, including preparing an org for installing packages, adding more customization after the package or application is deployed, and customizing specific orgs that are used in the product's development process.

Learn more about managing unpackaged metadata in [TODO: link Managing unpackaged configuration].

Project Orgs & Services
-----------------------

Orgs and services are external, authenticated resources that each project uses. CumulusCI makes it easy to connect orgs and services to a single project, or to use them across many projects.

Orgs
^^^^

> Each project has its own set of orgs, including active scratch orgs, persistent orgs like a production or packaging org, and predefined scratch org configurations. CumulusCI securely stores org authentication information in its keychain (via a unique identifier of characters), making it easy to access connected orgs at any time. The ``cci org list`` command shows all of the orgs connected to a project. Orgs can also be shared across multiple projects.

Reviewer 3: ```suggestion
Each project has its own set of orgs, including active scratch orgs, persistent orgs such as a production or packaging org, and predefined scratch org configurations. CumulusCI securely stores org authentication information in its keychain (via a unique ID), making it easy to access connected orgs at any time. The ``cci org list`` command shows all of the orgs connected to a project. Orgs can be shared across multiple projects.
```
"unique identifier of characters" sounds a bit awkward. Trim to "unique ID" unless it's important to specify that the ID is character-based?

Reviewer: Everyone hates describing a keychain. (Although to be fair, it's pretty easy to intuit.) 

Glick and Reed both suggested deleting mention of this. It's gone. :)

> Each project has its own set of orgs, including active scratch orgs, persistent orgs like a production or packaging org, and predefined scratch org configurations. CumulusCI securely stores org authentication information in its keychain (via a unique identifier of characters), making it easy to access connected orgs at any time. The ``cci org list`` command shows all of the orgs connected to a project. Orgs can also be shared across multiple projects.

Reviewer: You folks really hate describing what a keychain is. (Glick suggested I erase the entire paragraph for it in the Project Structure section.)

> Configuring orgs in CumulusCI is powerful, but comes with some complexity. To review the details, read the sections [TODO: link Scratch org environments] and [TODO: link persistent org section].

Reviewer 3: ```suggestion
Configuring orgs in CumulusCI is powerful, but comes with some complexity. For details, see [TODO: link Scratch org environments] and [TODO: link persistent org section].
```

Services
^^^^^^^^

Services represent external resources used by CumulusCI automation, such as access to a GitHub account or a MetaDeploy instance. Services are usually, but not always, connected to CumulusCI across projects as part of the global keychain. The command ``cci service list`` shows you which services are connected in the context of the current project.

> Global services are easy to use and share and, therefore, we recommend you rely on them for the most part. However, services can also be connected at the project level, which means that they're scoped to a single project and cannot be shared.

Reviewer 3: ```suggestion
Global services are easy to use and share, and we recommend that you use them when possible. However, services can also be connected at the project level, which means that they're scoped to a single project and can't be shared.
```

Reviewer: "Global services are easy to use and share; we recommend that you use them as much as possible."

More conversational in tone. And I'm still gonna advocate for the "cannot" here for emphasis.

Reviewer 3: Sounds good! That semicolon seems fussy for such a short sentence. Replace with a period?

Roger on "cannot."

>         Example: If you need to use a specific Dev Hub for one - and only one(!) - project, you can connect to that service with the command ``cci service connect devhub --project``.

Reviewer 3: Replace with em dashes.

Reviewer: So it's looking like that's not feasible with Sphinx. And one small dash is too short. Can we use two small dashes instead? (Any command with two dashes is in that special ``code`` formatting.)

Reviewer 3: Yep, let's do two small en dashes, then. If we ever convert this content to easyDITA, we can do a quick search and replace.

>         Example: If you need to use a specific Dev Hub for one - and only one(!) - project, you can connect to that service with the command ``cci service connect devhub --project``.

Reviewer: I left this in because it was in the original text, and an example never hurts (even if, yes, it's a bit detailed).

@Br4nd0R & @davisagli: Thoughts on excising this?

>         Example: If you need to use a specific Dev Hub for one - and only one(!) - project, you can connect to that service with the command ``cci service connect devhub --project``.

Reviewer 2: No strong opinion either way, but I'll note that we also cover this in a subsection of the topic on scratch orgs: https://cumulusci.readthedocs.io/en/feature-doc-reorg__scratch-org/scratch_orgs.html#use-a-non-default-dev-hub

Tasks and Flows
---------------

CumulusCI uses a framework of **tasks** and **flows** to organize the automation that is available to each project.

> Tasks are units of automation. A task could perform a deployment, load a dataset, retrieve material from an org, install a managed package, or undertake many other activities. CumulusCI ships with scores of tasks out of the box.

Reviewer 3: ```suggestion
Tasks are units of automation. A task can perform a deployment, load a dataset, retrieve data from an org, install a managed package, or do many other things. CumulusCI ships with scores of tasks out of the box.
```

> Popular tasks:

Reviewer 3: ```suggestion
Popular task commands include:
```

* ``cci task list``: Review the tasks available in a project.
* ``cci task info <name>``: Learn more about a task and how to configure its options; ``<name>`` is the name of the task.

> * ``cci task run <name> --org <org>``: Run a task; ``<name>`` is the name of the task and ``<org>`` is the org you'd like to run it against.

Reviewer 3: ```suggestion
* ``cci task run <name> --org <org>``: Run a task; ``<name>`` is the name of the task and ``<org>`` is the org to run it against.
```

        Example: ``run_tests`` executes Apex unit tests. If you have an org called ``dev``, you can run this task against it with the command ``cci task run run_tests --org dev``

> Many operations that you'll undertake with CumulusCI, including creating new orgs, use flows. Flows are ordered sequences of tasks (and other flows!) that produce a cohesive outcome, such as an org that's configured to suit a specific workflow.

Reviewer 3: ```suggestion
Many operations in CumulusCI, including creating new orgs, use flows. Flows are ordered sequences of tasks (and even other flows) that produce a cohesive outcome, such as an org that's configured to suit a specific workflow.
```

> Popular flows:

Reviewer 3: ```suggestion
Popular flow commands include:
```

> Popular flows:

Reviewer: @davidmreed Was going to change it to "popular flow commands" per Yee's suggestion. Are we being _too_ particular with flow-related? If it's important to discern, I have no problem putting it in. 

And to confirm: If we go with flow-related, do we also say "task-related commands" with the suggestions in the previous section? Just want to be accurate.

* ``cci flow list``: Review the flows available in a project.

> * ``cci flow info <name>``: Learn more about a flow and the tasks it contains; ``<name>`` is the name of the flow

Reviewer 3: ```suggestion
* ``cci flow info <name>``: Learn more about a flow and the tasks it contains; ``<name>`` is the name of the flow.
```

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer 3: ```suggestion
* ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.
```
```suggestion
* ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org to run it against.
```

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer: I am severely resistant to having it all lowercase. This is technically its own sentence, not an extension of one. 

Does it help to note that the reason why I put in colons was to save writing "allows user to" each time, which looks a bit goofy when repeated in a list. We can default to "reviews/learns/runs", but "learns more about a flow" or "runs a flow" doesn't hit my ear right when read aloud.

@lianeyee, thoughts?

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer 4: How about 

```suggestion
* ``cci flow run <name> --org <org>``: Run the flow ``<name>`` against the org ``<org>``.
```

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer: It's more succinct, but now it might not also be as clear to the reader. (This is where we start getting into the debate of who are we writing this for, and how much do they know or can intuit.)

Can we get more of the team to chime in on this one? @davisagli? @Br4nd0R?

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer 3: @justinvholt I agree with using the colon to economize on words. (Another option we sometimes use for such lists is an em dash, but I wouldn't here because a dash seems slightly less clear where the command stops and the separator starts.) 

For me, it is clearer to see the info spelled out in two separate parts. "Here's what the flow does. Here are the variables you specify." How about replacing the semicolons with periods? Seems simpler. And use ending periods on all. 

For extra succinctness, omit "you'd like."

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Author: I like the more suggestion by @davidmreed. True, it may not be as clear, but including an example (as is present just after the 'popular tasks' list) should clear that up.

> * ``cci flow run <name> --org <org>``: Run a flow; ``<name>`` is the name of the flow and ``<org>`` is the org you'd like to run it against.

Reviewer: @Br4nd0R: What example would you like to insert?

> Many of the most common flows you'll work with in CumulusCI are designed to build and configure specific orgs for you. Here's a few of the most common flows that build orgs:

Reviewer 3: ```suggestion
Many of the most common flows in CumulusCI are designed to build and configure specific orgs for you. Here's a few of the most common flows that build orgs.
```

* ``dev_org``: This is an unmanaged org that is designed for development use. This flow is typically used with an org whose configuration is ``dev`` or ``dev_namespaced``.
* ``qa_org``: This is an unmanaged org that is designed for testing. This flow is typically used with an org whose configuration is ``qa``.
* ``install_beta``: This is a managed org with the latest beta release installed, for projects that build managed packages. This flow is typically used with an org whose configuration is ``beta``.
* ``install_prod``: This is a managed org with the latest release installed, for projects that build managed packages.
* ``regression_org``: This is a managed org that starts with the latest release installed and is then upgraded to the latest beta to simulate a subscriber upgrade for projects that build managed packages. This flow is typically used with an org whose configuration is ``release``.

> CumulusCI derives the library of tasks and flows available for any project by combining its internal standard library with your customizations in ``cumulusci.yml``. Customizations can add new tasks and flows, customize the way tasks behave, and extend, combine, and modify flows to suit the specific needs of the project. We cover customization in depth in [TODO: reference Customizing CumulusCI].

Reviewer 3: ```suggestion
CumulusCI derives the library of tasks and flows available for any project by combining its internal standard library with your customizations in ``cumulusci.yml``. Customizations can add new tasks and flows, customize the way tasks behave, and extend, combine, and modify flows to better suit the project's needs. We cover customization in depth in [TODO: reference Customizing CumulusCI].
```

> * ``dev_org``: This is an unmanaged org that is designed for development use. This flow is typically used with an org whose configuration is ``dev`` or ``dev_namespaced``.

Reviewer: "Here’s a few of the most common flows that build orgs:"

Just noticed that this is a bit confusing when I read it aloud. So to confirm: this is the name of a flow itself, not just the org? Both? Should we say "this flow creates a(n) _______ org"?

Reviewer 2: These are the names of flows. When running the flow you also have to specify which org config to run it against.

Reviewer: So would it be better to say "This flow builds/creates/heroically unleashes a ______ org blah blah blah"?

Reviewer 3: If we go with "This flow builds/creates/etc." (which sounds good to me), consider starting the second sentences with "It's typically used..."

Reviewer: Soooooooooooooo satisfying to fix this paragraph.

* ``qa_org``: This is an unmanaged org that is designed for testing. This flow is typically used with an org whose configuration is ``qa``.
* ``install_beta``: This is a managed org with the latest beta release installed, for projects that build managed packages. This flow is typically used with an org whose configuration is ``beta``.
* ``install_prod``: This is a managed org with the latest release installed, for projects that build managed packages.
* ``regression_org``: This is a managed org that starts with the latest release installed and is then upgraded to the latest beta to simulate a subscriber upgrade for projects that build managed packages. This flow is typically used with an org whose configuration is ``release``.

> CumulusCI derives the library of flows and tasks available for any project by combining its internal standard library with your customizations in ``cumulusci.yml``. Customizations can add new tasks and flows, customize the way featured tasks behave, and extend, combine, and modify featured flows to suit the specific needs of the project. We cover customization in depth in [TODO: reference Customizing CumulusCI].

Reviewer 2: I'm not understanding your use of "featured" here.

Reviewer: @davisagli I don't like _out-of-the-box_, thought _featured_ would be a better substitute. I can change it back, or is there a better word?

Reviewer 2: out-of-the-box and featured don't seem like synonyms to me. The point we're trying to make is that a project can either use the standard tasks and flows as is, or customize them, or add new ones. Any of the standard tasks and flows can be customized, not just "featured" ones.

Reviewer: Then we won't use either, simply go with tasks.
